### PR TITLE
#141 add simple product to cart

### PR DESCRIPTION
**Pull request by roma-glushko**
### Description
The pull enables adding simple products to customer/guest carts via GraphQl mutations.

In order to add simple product to cart the following information can be passed:
- product SKUs (required)
- quantities (required)
- custom options (required if product has required options)

As a result of executing such mutation, the current state of the cart will be returned. Cart items information can vary and depends on product types. For this case, simple cart item contains information about custom options.

### Fixed Issues
1. magento/magento2#141: Add simple product to cart

### Manual testing scenarios
The following query can be executed to test the changes:

```
mutation {
  addSimpleProductsToCart(
    input: {
      cart_id: "b96566a020b8797b7d806a948174275e", 
      cartItems: [
        {
          data: {
            sku: "product_dynamic_24494", 
            qty: 1
          }, 
          customizable_options: [
            {id: 5, value: "test"},
            {id: 6, value: "5"}
          ]
        }
      ]
    }
  ) {
    
    cart {
      items {
        id
        qty
        ... on SimpleCartItem {
          customizable_options {
            id
            label
            type
            is_required
            values {
              id
              label
              value
              price {
                type
                units
                value
              }
            }
          }
        }
        product {
          sku
          name
          image
          url_key
          categories {
            id
            name
          } 
          updated_at
          websites {
            name
          }
        }
      }
    }    
  }
}
```

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=170) <br/>All committers have signed the CLA.

File: app/code/Magento/CatalogGraphQl/README.md
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/README.md]: Looks like accidental change

File: app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php:17]: Why is this one needed?

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php
Comment (paliarush) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php]: I believe this part, including files processing will be completed before we merge.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemTypeResolverComposite.php
Comment (paliarush) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemTypeResolverComposite.php]: I don't think this error will be useful for the end user. This should probably be a regular `\LogicException`. Same for the other exception below.

 What do you think?

File: app/code/Magento/QuoteGraphQl/etc/schema.graphqls
Comment (paliarush) [inline app/code/Magento/QuoteGraphQl/etc/schema.graphqls]: Why ID can be null?

File: app/code/Magento/SimpleProductGraphQl/Model/Resolver/Cart/AddSimpleProductsToCart.php
Comment (paliarush) [inline app/code/Magento/SimpleProductGraphQl/Model/Resolver/Cart/AddSimpleProductsToCart.php]: It is probably better to use semicolon or another separator for the messages in GraphQL output.

File: app/code/Magento/SimpleProductGraphQl/composer.json
Comment (paliarush) [inline app/code/Magento/SimpleProductGraphQl/composer.json]: This should be part of `CatalogGraphql` module since simple product type is defined in `Catalog` module.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/SimpleProduct/AddSimpleProductsToCartTest.php
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/SimpleProduct/AddSimpleProductsToCartTest.php]: We should not have `var_dump` in tests.

File: dev/tests/api-functional/testsuite/Magento/GraphQl/SimpleProduct/AddSimpleProductsToCartTest.php
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/SimpleProduct/AddSimpleProductsToCartTest.php]: Would be perfect to validate all the requested fields.

**Review by paliarush — CHANGES_REQUESTED** [review]
Please address review comments above.

File: app/code/Magento/CatalogGraphQl/README.md
Comment (roma-glushko) [reply app/code/Magento/CatalogGraphQl/README.md]: Indeed. Will fix that. Thank you for noticing!

File: app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php
Comment (roma-glushko) [reply app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php:17]: The reason why it's needed is ProductDateOptionType wants to see date value as an associate array (like ['year' => '2018', 'month' => '1', 'day' => '1']). By this changes, we allow to use string as a date value. This meets our schema shape. 
The same approach is used in REST API.

File: app/code/Magento/QuoteGraphQl/etc/schema.graphqls
Comment (roma-glushko) [reply app/code/Magento/QuoteGraphQl/etc/schema.graphqls]: It should not be. Changing to `id: Int!`. Thank you for mentioning!

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php
Comment (roma-glushko) [reply app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php]: Yes, I have been implemented this mock. Thank you for noticing!

**Comment by naydav** [comment]
@roma-glushko 
Pls, merge mainline and re-run tests

I found a few problems

1)
```

FILE: ...pp/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php
----------------------------------------------------------------------
FOUND 6 ERRORS AFFECTING 6 LINES
----------------------------------------------------------------------
 19 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 21 | ERROR | [x] {@inheritdoc} does not import parameter annotation
 30 | ERROR | [x] Missing short description
 31 | ERROR | [x] There must be exactly one blank line before tags
 53 | ERROR | [x] Missing short description
 54 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
PHPCBF CAN FIX THE 6 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ml/app/code/Magento/QuoteGraphQl/Model/Hydrator/CartHydrator.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 15 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 19 | ERROR | [x] Missing short description
 20 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...nto/QuoteGraphQl/Model/Resolver/Cart/AddSimpleProductsToCart.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
  32 | ERROR | [x] {@inheritdoc} imports only short description,
     |       |     annotation must have long description
 115 | ERROR | [x] {@inheritdoc} imports only short description,
     |       |     annotation must have long description
 117 | ERROR | [x] {@inheritdoc} does not import parameter annotation
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...nto/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php
----------------------------------------------------------------------
FOUND 8 ERRORS AFFECTING 8 LINES
----------------------------------------------------------------------
  29 | ERROR | [x] {@inheritdoc} imports only short description,
     |       |     annotation must have long description
  64 | ERROR | [x] {@inheritdoc} imports only short description,
     |       |     annotation must have long description
  66 | ERROR | [x] {@inheritdoc} does not import parameter annotation
 104 | ERROR | [x] Missing short description
 105 | ERROR | [x] There must be exactly one blank line before tags
 209 | ERROR | [x] Missing short description
 210 | ERROR | [x] There must be exactly one blank line before tags
 225 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
PHPCBF CAN FIX THE 8 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ode/Magento/QuoteGraphQl/Model/Resolver/CartItemTypeResolver.php
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 4 LINES
----------------------------------------------------------------------
 14 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 32 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 33 | ERROR | [x] There must be exactly one blank line before tags
 33 | ERROR | [x] There must be exactly one blank line between lines
 35 | ERROR | [x] {@inheritdoc} does not import parameter annotation
----------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...to/QuoteGraphQl/Model/Resolver/CartItemTypeResolverComposite.php
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 4 LINES
----------------------------------------------------------------------
 14 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 32 | ERROR | [x] {@inheritdoc} imports only short description,
    |       |     annotation must have long description
 33 | ERROR | [x] There must be exactly one blank line before tags
 33 | ERROR | [x] There must be exactly one blank line between lines
 35 | ERROR | [x] {@inheritdoc} does not import parameter annotation
----------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```
2)
```
/var/www/html/app/code/Magento/QuoteGraphQl/Model/Resolver/Cart/AddSimpleProductsToCart.php:34	The class AddSimpleProductsToCart has a coupling between objects value of 16. Consider to reduce the number of dependencies under 13.
/var/www/html/app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php:111	The method getOptionData() has a Cyclomatic Complexity of 15. The configured cyclomatic complexity threshold is 10.
/var/www/html/app/code/Magento/QuoteGraphQl/Model/Resolver/CartItem/CustomizableOptions.php:111	The method getOptionData() has an NPath complexity of 720. The configured NPath complexity threshold is 200.
```

**Review by paliarush — APPROVED** [review]

**Review by paliarush — APPROVED** [review]

File: app/code/Magento/CatalogGraphQl/etc/graphql/di.xml
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/etc/graphql/di.xml:9]: This looks risky because we override the default behavior of catalog class. Why is this needed?

Comment (roma-glushko) [reply app/code/Magento/CatalogGraphQl/etc/graphql/di.xml:9]: @paliarush It's needed for API purpose only (that's why it's defined in scope of graphql area) to bypass requirement for date values which Magento has on the frontend (date value should be an array). So by this, we made it possible to receive and validate data value in string format. 

Does it make sense?

File: app/code/Magento/QuoteGraphQl/Model/Resolver/DataProvider/Cart/CartHydrator.php
Comment (rogyar) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/DataProvider/Cart/CartHydrator.php]: In terms of classic meaning of hydrator I would name this method as `extract`. The `hydrate` makes the opposite action: casts array to an object

Comment (roma-glushko) [reply app/code/Magento/QuoteGraphQl/Model/Resolver/DataProvider/Cart/CartHydrator.php]: Agreed 👍 It should a data provider

**Pull request merged**
